### PR TITLE
Move reachability logic to CustomNavigationController

### PR DESCRIPTION
**Pull request by haiyingweng**
## Overview
`HomeMapVC` and `RouteOptionsVC` use very similar reachability logic. This PR moves the reachability logic and banner for "No internet connection" to `CustomNavController` 
 

## Changes Made
- Moved reachability logic and banner for no internet connection in `HomeMapVC` and `RouteOptionsVC` to `CustomNavController`
- Added delegate protocol to enable/disable user interaction depending on reachability
- Made sure reachability delegate is set properly every time a ViewController that conform to the protocol is pushed/popped 

## Test Coverage
Connected/Disconnected to wifi a bunch of time and banner for "no internet connection" is dismissed/displayed as wanted. 

## Next Steps 
~~`RouteOptionsVC` still manages banners for route calculation problems (e.g. out of range, cannot connect to server). We could possibly also move those to `CustomNavController`~~

## Related PRs or Issues 
#274

### CODE REVIEW

File: TCAT/Controllers/CustomNavigationController.swift [hunk 1]
```diff
@@ -23,6 +28,9 @@ class CustomNavigationController: UINavigationController, UINavigationController
         .foregroundColor: Colors.black
     ]
 
+    private var banner = StatusBarNotificationBanner(title: Constants.Banner.noInternetConnection, style: .danger)
+    private let reachability = Reachability()!
```
Comment (kevinchan159) [inline TCAT/Controllers/CustomNavigationController.swift]: delete `!` at end of line

File: TCAT/Controllers/CustomNavigationController.swift [hunk 2]
```diff
@@ -51,13 +63,31 @@ class CustomNavigationController: UINavigationController, UINavigationController
             Analytics.shared.log(payload)
         }
     }
+    
+    override func viewWillAppear(_ animated: Bool) {
+
```
Comment (kevinchan159) [inline TCAT/Controllers/CustomNavigationController.swift:69]: delete newline

File: TCAT/Controllers/CustomNavigationController.swift [hunk 3]
```diff
@@ -51,13 +63,31 @@ class CustomNavigationController: UINavigationController, UINavigationController
             Analytics.shared.log(payload)
         }
     }
+    
+    override func viewWillAppear(_ animated: Bool) {
```
Comment (kevinchan159) [inline TCAT/Controllers/CustomNavigationController.swift:67]: you need to call `super.viewDidAppear()` as well

File: TCAT/Controllers/CustomNavigationController.swift [hunk 4]
```diff
@@ -99,6 +129,22 @@ class CustomNavigationController: UINavigationController, UINavigationController
     @objc private func backAction() {
         _ = popViewController(animated: true)
     }
+    
+    @objc func reachabilityChanged(notif: Notification) {
+
```
Comment (kevinchan159) [inline TCAT/Controllers/CustomNavigationController.swift]: delete newline

File: TCAT/Controllers/HomeMapViewController.swift [hunk 5]
```diff
@@ -239,6 +193,12 @@ extension HomeMapViewController: HomeOptionsCardDelegate {
     }
 }
 
+extension HomeMapViewController: ReachabilityDelegate {
+    func reachabilityChanged(connection: Reachability.Connection) {
+        delegate?.reachabilityChanged(connection: connection)
+    }
+}
```
Comment (kevinchan159) [inline TCAT/Controllers/HomeMapViewController.swift]: newline after 196 and before 200

File: TCAT/Controllers/RouteOptionsViewController+Extensions.swift [hunk 6]
```diff
@@ -450,3 +450,14 @@ extension RouteOptionsViewController: RouteSelectionViewDelegate {
...
+extension RouteOptionsViewController: ReachabilityDelegate {
+    func reachabilityChanged(connection: Reachability.Connection) {
+        switch connection {
+        case .none:
+            setUserInteraction(to: false)
+        case .cellular, .wifi:
+            setUserInteraction(to: true)
+        }
+    }
+}
```
Comment (kevinchan159) [inline TCAT/Controllers/RouteOptionsViewController+Extensions.swift]: new line after 454 and before 463

File: TCAT/Controllers/CustomNavigationController.swift [hunk 7]
```diff
@@ -23,6 +28,9 @@ class CustomNavigationController: UINavigationController, UINavigationController
         .foregroundColor: Colors.black
     ]
 
+    private var banner = StatusBarNotificationBanner(title: Constants.Banner.noInternetConnection, style: .danger)
```
Comment (kevinchan159) [inline TCAT/Controllers/CustomNavigationController.swift]: is this ever changed? if not can we make it `private let`

File: TCAT/Controllers/CustomNavigationController.swift [hunk 8]
```diff
@@ -51,13 +63,34 @@ class CustomNavigationController: UINavigationController, UINavigationController
...
+        // Add Notification Observers
+        NotificationCenter.default.addObserver(self,
+                                               selector: #selector(reachabilityChanged(notif:)),
+                                               name: .reachabilityChanged,
+                                               object: reachability)
+        guard let reach = reachability else {return}
+        do {
+            try reach.startNotifier()
+        } catch {
+            print("could not start reachability notifier")
```
Comment (lucyxubroad) [inline TCAT/Controllers/CustomNavigationController.swift]: Are we going to use the "printClass"? If not, update the print message (correct punctuation).

**Review by lucyxubroad — CHANGES_REQUESTED** [review]

File: TCAT/Controllers/CustomNavigationController.swift [hunk 9]
```diff
@@ -119,16 +167,32 @@ class CustomNavigationController: UINavigationController, UINavigationController
                 viewController.navigationItem.setLeftBarButton(customBackButton(), animated: true)
             }
         }
+        
+        if let homeMapVC = viewController as? HomeMapViewController {
+            reachabilityDelegate = homeMapVC
+        }
+        if let routeOptionsVC = viewController as? RouteOptionsViewController {
```
Comment (lucyxubroad) [inline TCAT/Controllers/CustomNavigationController.swift]: Is this supposed to be "if, else if"?

File: TCAT/Controllers/CustomNavigationController.swift [hunk 10]
```diff
@@ -119,16 +167,32 @@ class CustomNavigationController: UINavigationController, UINavigationController
...
     override func popViewController(animated: Bool) -> UIViewController? {
 
         let viewController = super.popViewController(animated: animated)
-        if let lastViewController = viewControllers.last as? HomeMapViewController {
-            lastViewController.navigationItem.leftBarButtonItem = nil
+        if let homeMapVC = viewControllers.last as? HomeMapViewController {
+            homeMapVC.navigationItem.leftBarButtonItem = nil
+            reachabilityDelegate = homeMapVC
+        }
+        if let routeOptionsVC = viewControllers.last as? RouteOptionsViewController {
```
Comment (lucyxubroad) [inline TCAT/Controllers/CustomNavigationController.swift]: Is this "if, else if"?

File: TCAT/Controllers/RouteOptionsViewController.swift [hunk 11]
```diff
@@ -130,21 +129,17 @@ class RouteOptionsViewController: UIViewController {
...
-        // Takedown reachability
-        reachability?.stopNotifier()
-        NotificationCenter.default.removeObserver(self, name: .reachabilityChanged, object: reachability)
         // Remove banner
         banner?.dismiss()
         banner = nil
         routeTimer?.invalidate()
         updateTimer?.invalidate()
-        // Stop observing when app becomes active 
+        // Stop observing when app becomes active
```
Comment (danielVebman) [inline TCAT/Controllers/RouteOptionsViewController.swift]: I'm a little confused by this comment. Doesn't the next line stop observing when the view controller is dismissed?

File: TCAT/Controllers/CustomNavigationController.swift [hunk 12]
```diff
@@ -51,13 +63,34 @@ class CustomNavigationController: UINavigationController, UINavigationController
...
 
     override func viewWillDisappear(_ animated: Bool) {
         super.viewWillDisappear(animated)
 
         if let screenshotObserver = screenshotObserver {
             NotificationCenter.default.removeObserver(screenshotObserver)
         }
+        
+        // Remove Notification Observers
+        guard let reach = reachability else {return}
```
Comment (kevinchan159) [inline TCAT/Controllers/CustomNavigationController.swift]: same comment about `reachability` and also add spaces: `{ return }`

File: TCAT/Controllers/CustomNavigationController.swift [hunk 13]
```diff
@@ -51,13 +63,34 @@ class CustomNavigationController: UINavigationController, UINavigationController
...
+    
+    override func viewWillAppear(_ animated: Bool) {
+        super.viewWillAppear(animated)
+        
+        // Add Notification Observers
+        NotificationCenter.default.addObserver(self,
+                                               selector: #selector(reachabilityChanged(notif:)),
+                                               name: .reachabilityChanged,
+                                               object: reachability)
+        guard let reach = reachability else {return}
```
Comment (kevinchan159) [inline TCAT/Controllers/CustomNavigationController.swift]: can we do `guard let reachability = reachability` just so theres no confusion as to what `reach` is

File: TCAT/Controllers/RouteOptionsViewController+Extensions.swift [hunk 14]
```diff
@@ -450,3 +450,16 @@ extension RouteOptionsViewController: RouteSelectionViewDelegate {
...
+
+extension RouteOptionsViewController: ReachabilityDelegate {
+    
+    func reachabilityChanged(connection: Reachability.Connection) {
+        switch connection {
+        case .none:
+            setUserInteraction(to: false)
+        case .cellular, .wifi:
+            setUserInteraction(to: true)
+        }
```
Comment (kevinchan159) [inline TCAT/Controllers/RouteOptionsViewController+Extensions.swift]: could we just do `setUserInteraction(to: connection != .none)`

File: TCAT/Controllers/RouteOptionsViewController.swift (hunk 11 above)
Comment (kevinchan159) [reply TCAT/Controllers/RouteOptionsViewController.swift]: +1

Comment (haiyingweng) [reply TCAT/Controllers/RouteOptionsViewController.swift]: Yea, I agree. The comment doesn't make sense. I think the person that put in the comment meant to say stop observing for the notification called `appBecameActiveNotification`. But will fix that!

File: TCAT/Controllers/CustomNavigationController.swift [hunk 15]
```diff
@@ -99,6 +132,21 @@ class CustomNavigationController: UINavigationController, UINavigationController
...
+    
+    @objc func reachabilityChanged(notif: Notification) {
+        if let reachability = notif.object as? Reachability {
+            switch reachability.connection {
+            case .wifi, .cellular:
+                banner.dismiss()
+            case .none:
+                banner.show(queuePosition: .front, on: self)
+                banner.autoDismiss = false
+                banner.isUserInteractionEnabled = false
```
Comment (Omarrasheed) [inline TCAT/Controllers/CustomNavigationController.swift:150]: Ik this was from previous code but can we take this line out? I don't think we should make this banner not dismissible.

File: TCAT/Controllers/CustomNavigationController.swift [hunk 16]
```diff
@@ -23,6 +28,9 @@ class CustomNavigationController: UINavigationController, UINavigationController
         .foregroundColor: Colors.black
     ]
 
+    private let banner = StatusBarNotificationBanner(title: Constants.Banner.noInternetConnection, style: .danger)
+    private let reachability = Reachability()
```
Comment (Omarrasheed) [inline TCAT/Controllers/CustomNavigationController.swift:32]: Seems like we're only ever accessing this in the `viewWillAppear` function to add it as the object being passed into the notification observer and then we safely unwrap in our `reachabilityChanged` function. Could we look into moving this into the `viewWillAppear` function as a local variable instead? Not 100% about memory-related implications for this object tho

File: TCAT/Controllers/CustomNavigationController.swift [hunk 17]
```diff
@@ -119,16 +167,30 @@ class CustomNavigationController: UINavigationController, UINavigationController
                 viewController.navigationItem.setLeftBarButton(customBackButton(), animated: true)
             }
         }
+        
+        if let homeMapVC = viewController as? HomeMapViewController {
+            reachabilityDelegate = homeMapVC
+        } else if let routeOptionsVC = viewController as? RouteOptionsViewController {
+            reachabilityDelegate = routeOptionsVC
+        }
```
Comment (Omarrasheed) [inline TCAT/Controllers/CustomNavigationController.swift]: Do we want this to only work for certain viewControllers or is this simply a check to see if the viewController conforms to the protocol? If its the latter, then we can change this to be 
```swift
if let vc = viewController as? ReachabilityDelegate {
  ...
}
```

File: TCAT/Controllers/CustomNavigationController.swift [hunk 18]
```diff
@@ -119,16 +167,30 @@ class CustomNavigationController: UINavigationController, UINavigationController
...
     override func popViewController(animated: Bool) -> UIViewController? {
 
         let viewController = super.popViewController(animated: animated)
-        if let lastViewController = viewControllers.last as? HomeMapViewController {
-            lastViewController.navigationItem.leftBarButtonItem = nil
+        if let homeMapVC = viewControllers.last as? HomeMapViewController {
+            homeMapVC.navigationItem.leftBarButtonItem = nil
+            reachabilityDelegate = homeMapVC
+        } else if let routeOptionsVC = viewControllers.last as? RouteOptionsViewController {
+            reachabilityDelegate = routeOptionsVC
```
Comment (Omarrasheed) [inline TCAT/Controllers/CustomNavigationController.swift]: Same as previous comment

**Review by Omarrasheed — APPROVED** [review]
Looks great!

File: TCAT/Controllers/CustomNavigationController.swift (hunk 15 above)
Comment (haiyingweng) [reply TCAT/Controllers/CustomNavigationController.swift:150]: It makes sense to me to not make the banner dismissible by the user though. I feel we don't want the user to be able to dismiss the banner bc they might get confused if they dismissed it and can't distinguish between when they have internet or not.

File: TCAT/Controllers/CustomNavigationController.swift (hunk 16 above)
Comment (haiyingweng) [reply TCAT/Controllers/CustomNavigationController.swift:32]: but we're also accessing `reachability` in `viewWillDisappear` to stop notifier

File: TCAT/Controllers/CustomNavigationController.swift (hunk 12 above)
Comment (kevinchan159) [reply TCAT/Controllers/CustomNavigationController.swift]: bump regarding spaces before and after `return`

**Review by kevinchan159 — APPROVED** [review]
🚢

**Review by lucyxubroad — APPROVED** [review]
looks great!

**Pull request merged**